### PR TITLE
Use a lister for Prowjobs in sinker

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR changes sinker to use a lister for prowjobs. This is done by using the dynamic client from controller-runtime. 

The implementation that is used here has a `Reader` that is backed by a dynamic cache, meaning whenever any object is requested through it, a cache/lister for that object type will be created. The only exception is `unstructured.Unstructured`.

A couple of other things were changed along the way:

* Sinker was split into a `cmd/sinker` and a `sinker` package to more cleanly separate the actual control loop from the initialization
* The kubernetes dependecies were upgraded to Kubernetes v1.13, this is required for controller-runtime
* Sinker now uses leader election
* The `build` controller was changed to use the controller-runtime client, because it had a dependency on a vendored, typed client that doesn't build anymore with the new client-go version. Tests pass, but I can not test if it actually works.

The actual business logic and the tests of sinker were not changed, just split among the two packages. Unfortunately Gihub does not recognize this.

/assign @fejta @stevekuznetsov @cjwagner

### CODE REVIEW

File: prow/cmd/sinker/main.go [hunk 1]
```diff
@@ -109,9 +98,9 @@ func main() {
...
 	metrics.ExposeMetrics("sinker", pushGateway.Endpoint, pushGateway.Interval.Duration)
 
-	prowJobClient, err := o.kubernetes.ProwJobClient(cfg().ProwJobNamespace, o.dryRun.Value)
+	controlClusterCfg, err := o.kubernetes.ControlClusterConfig(o.dryRun.Value)
 	if err != nil {
-		logrus.WithError(err).Fatal("Error creating ProwJob client.")
+		logrus.WithError(err).Fatal("Error getting controlcluster config.")
 	}
 
 	buildClusterClients, err := o.kubernetes.BuildClusterClients(cfg().PodNamespace, o.dryRun.Value)
```
Comment (alvaroaleman) [inline prow/cmd/sinker/main.go]: After adding special handling and tests to ensure that sinker will only log messages and not do apicalls to Prowjobs when `dryRun` is enabled, I ralized that enabling that will make sinker crash here, as `BuildClusterClients` does not support it. Should I remove the code that handles `dryRun` again?

**Comment by alvaroaleman** [comment]
The test failure is because the vendored knative clientset is incompatible with the new client-go version. We could:

* Try to find a compatible version of knative and do this every time we or knative changes the used client-go version
* Generate our own clientset for knative inside this repo
* Use the dynamic client from controller-runtime when acting on knative resources and get rid of the typed knative clientset entirely

Any preference?

File: prow/cmd/sinker/main.go (hunk 1 above)
Comment (stevekuznetsov) [reply prow/cmd/sinker/main.go]: I think we need to back away from the "DRY" `dryRun` behavior, it was applied too broadly and is not useful as you note.

**Comment by stevekuznetsov** [comment]
If @fejta is OK with the dynamic client for knative, I think it's the best way forward or we end up in dependency hell every time anyone changes anything.

**Comment by fejta** [comment]
Basically it looks like you're deleting and rewriting sinker. Can you make it more obvious which parts of sinker are getting reused? Ideally maximize the amount of reuse

**Comment by alvaroaleman** [comment]
> Everything happening here seems fine, but there is too much stuff happening in a single PR to be able to review coherently.

Valid, I will split this up and create distinct and smaller PRs

**Pull request closed without merging**